### PR TITLE
[PLT-711] Changed currency and index for a name to a List instead of a single value.

### PR DESCRIPTION
**Pull request by marc-opengamma**
Old bug but could have an impact on some gamma computation. 
Required for some SIMM analysis (forward looking estimates(.

### CODE REVIEW

File: projects/OG-Analytics/src/main/java/com/opengamma/analytics/financial/provider/calculator/discounting/CrossGammaSingleCurveCalculator.java [hunk 1]
```diff
@@ -81,7 +81,7 @@ public DoubleMatrix2D calculateCrossGamma(final InstrumentDerivative instrument,
     ArgumentChecker.isTrue(multicurve.getAllNames().size() == 1,
         "provider should have only one curve for GammaSingleCurve computation");
     String name = multicurve.getAllNames().iterator().next();
-    Currency ccy = multicurve.getCurrencyForName(name);
+    Currency ccy = multicurve.getCurrencyForName(name).get(0);
```
Comment (jodastephen) [inline projects/OG-Analytics/src/main/java/com/opengamma/analytics/financial/provider/calculator/discounting/CrossGammaSingleCurveCalculator.java]: Can the list be empty?

File: projects/OG-Analytics/src/main/java/com/opengamma/analytics/financial/provider/description/interestrate/MulticurveProviderDiscount.java [hunk 2]
```diff
@@ -57,15 +58,15 @@
   /**
    * Map of curve names to currencies.
    */
-  private Map<String, Currency> _namesToCurrency;
+  private Map<String, List<Currency>> _namesToCurrency;
```
Comment (julianhowarth) [inline projects/OG-Analytics/src/main/java/com/opengamma/analytics/financial/provider/description/interestrate/MulticurveProviderDiscount.java]: Why is this a list? What does the ordering imply? Or should it be a Set?

Comment (julianhowarth) [inline projects/OG-Analytics/src/main/java/com/opengamma/analytics/financial/provider/description/interestrate/MulticurveProviderDiscount.java]: Related to the above, would a Multimap (Set or List) be more appropriate?

File: projects/OG-Analytics/src/main/java/com/opengamma/analytics/financial/provider/description/interestrate/MulticurveProviderDiscount.java [hunk 3]
```diff
@@ -57,15 +58,15 @@
...
   /**
    * Map of curve names to ibor indices.
    */
-  private Map<String, IborIndex> _namesToIborIndex;
+  private Map<String, List<IborIndex>> _namesToIborIndex;
```
Comment (julianhowarth) [inline projects/OG-Analytics/src/main/java/com/opengamma/analytics/financial/provider/description/interestrate/MulticurveProviderDiscount.java]: SetMultimap?

File: projects/OG-Analytics/src/main/java/com/opengamma/analytics/financial/provider/description/interestrate/MulticurveProviderDiscount.java [hunk 4]
```diff
@@ -57,15 +58,15 @@
...
   /**
    * Map of curve names to overnight indices.
    */
-  private Map<String, IndexON> _namesToONIndex;
+  private Map<String, List<IndexON>> _namesToONIndex;
```
Comment (julianhowarth) [inline projects/OG-Analytics/src/main/java/com/opengamma/analytics/financial/provider/description/interestrate/MulticurveProviderDiscount.java]: SetMultimap?

File: projects/OG-Analytics/src/main/java/com/opengamma/analytics/financial/provider/description/interestrate/MulticurveProviderDiscount.java [hunk 5]
```diff
@@ -250,34 +254,44 @@ public YieldAndDiscountCurve getCurve(final String name) {
...
+   * Gets the currencies of a named curve. 
+   * If there is no discounting curve with this name in this provider, returns empty list.
    * @param name The name of a curve
-   * @return The currency, null if not found
+   * @return The currencies.
    */
-  public Currency getCurrencyForName(final String name) {
+  public List<Currency> getCurrencyForName(final String name) {
+    List<Currency> list = _namesToCurrency.get(name);
+    if (list == null) {
```
Comment (julianhowarth) [inline projects/OG-Analytics/src/main/java/com/opengamma/analytics/financial/provider/description/interestrate/MulticurveProviderDiscount.java]: These null checks will disappear if using a multimap

File: projects/OG-Analytics/src/main/java/com/opengamma/analytics/financial/provider/calculator/discounting/CrossGammaSingleCurveCalculator.java (hunk 1 above)
Comment (julianhowarth) [inline projects/OG-Analytics/src/main/java/com/opengamma/analytics/financial/provider/calculator/discounting/CrossGammaSingleCurveCalculator.java]: If you do know there's only one, then ``Iterators.getOnlyElement(...)`` may be a better approach

Comment (marc-opengamma) [inline projects/OG-Analytics/src/main/java/com/opengamma/analytics/financial/provider/calculator/discounting/CrossGammaSingleCurveCalculator.java]: From a business perspective, there is always at least one currency and we have check that there is exactly one curve.

File: projects/OG-Analytics/src/main/java/com/opengamma/analytics/financial/provider/description/interestrate/MulticurveProviderDiscount.java (hunk 2 above)
Comment (marc-opengamma) [inline projects/OG-Analytics/src/main/java/com/opengamma/analytics/financial/provider/description/interestrate/MulticurveProviderDiscount.java]: Those objects are used to loop on curves in different cases. The order is important is several cases.

**Comment by marc-opengamma** [comment]
Replaced Map<String, List<>>  by ListMultimap.

**Comment by marc-opengamma** [comment]
Replaced Map<String, List<>>  by ListMultimap.
Finished the changes following review.

**Pull request merged**